### PR TITLE
perf: reduce Context code

### PR DESCRIPTION
**Pull request by yusukebe**
### The author should do the following, if applicable

- [ ] Add tests
- [ ] Run tests
- [ ] `bun run format:fix && bun run lint:fix` to format the code
- [ ] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code

### CODE REVIEW

**Comment by yusukebe** [comment]
Whether we merge this or not is a difficult thing. In the `c.header()`, using `c.res` to store the header values will introduce a significant decrease in performance when using `c.header()`. Hmm.

**Comment by yusukebe** [comment]
Hello @usualoma @sushichan044 !

What do you think of this? If this makes sense to you, can you review it?

File: src/context.ts [hunk 1]
```diff
@@ -634,65 +596,33 @@ export class Context<
...
-      if (typeof v === 'string') {
-        this.#headers.set(k, v)
-      } else {
-        this.#headers.delete(k)
-        for (const v2 of v) {
-          this.#headers.append(k, v2)
+    if (headers) {
+      for (const [key, value] of Object.entries(headers)) {
+        if (Array.isArray(value)) {
+          for (const v of value) responseHeaders.append(key, v)
```
Comment (usualoma) [inline src/context.ts]: I think this is necessary to ensure compatibility.

```suggestion
          responseHeaders.delete(key)
          for (const v of value) {
            responseHeaders.append(key, v)
          }
```

File: src/context.ts [hunk 2]
```diff
@@ -783,34 +700,24 @@ export class Context<
...
-          : this.#newResponse(html, arg)
-      })
-    }
-
-    return typeof arg === 'number'
-      ? this.#newResponse(html as string, arg, headers)
-      : this.#newResponse(html as string, arg)
+    const TEXT_HTML = 'text/html; charset=UTF-8'
+    const res = (html: string) =>
+      this.#newResponse(html, arg, setDefaultContentType(TEXT_HTML, headers))
```
Comment (usualoma) [inline src/context.ts]: I think it's okay to specify it directly, but do you need to put it in a variable first?

```suggestion
      this.#newResponse(html, arg, setDefaultContentType('text/html; charset=UTF-8', headers))
```

**Review by usualoma — COMMENTED** [review]
Great, I think it's wonderful work!

File: src/context.ts (hunk 1 above)
Comment (yusukebe) [reply src/context.ts]: I updated the code following the former implementation and added the test: https://github.com/honojs/hono/pull/4100/commits/a722d539ff04126fd06b75f4a4f988efc03e8779

File: src/context.ts (hunk 2 above)
Comment (yusukebe) [reply src/context.ts]: You are right! Fixed https://github.com/honojs/hono/pull/4100/commits/47bb3287b7296363086135ec15b13a654c5fb387

**Review by sushichan044 — APPROVED** [review]
I had the same concern as @usualoma .
Since it has already been corrected, I think it looks good now!

I did a quick check on my end, and it seemed to handle common corner cases without any issues.

**Comment by sushichan044** [comment]
@yusukebe 
As a nitpick, it might be good to have a test like this to demonstrate the behavior when we input empty array as header value.
This test isn't essential, as the behavior is predictable from the implementation.

```ts
  it('Should remove existing header when new value is empty array', async () => {
    c.header('X-Test', 'existing')
    const res = c.json({ test: 'data' }, 200, {
      'X-Test': [],
    })
    expect(res.headers.get('X-Test')).toBeNull()
  })
```

**Review by usualoma — APPROVED** [review]
Thank you!

**Comment by yusukebe** [comment]
@usualoma @sushichan044 Thanks!

@sushichan044 Yes. [The test you mentioned](https://github.com/honojs/hono/pull/4100#issuecomment-2952181025) is good. I added it.

I'll include this in the next minor release.
